### PR TITLE
5 implement simulation module

### PR DESCRIPTION
**Pull request by Nicolaad**
adds functionality to simulate rounds based on a list of attacking and defending units.

### CODE REVIEW

File: client/core/src/com/cows/game/roundSimulation/RoundSimulator.kt [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+        val path = Map.getPathCoordinates()
+
+        //init simulation
+        val units = mutableListOf<UnitSimulationModel>()
+        val towers = mutableListOf<TowerSimulationModel>()
+
+
+        var timeToNextSpawn = 0
+        attackInstruction.forEach{
```
Comment (AndreasWintherMoen) [inline client/core/src/com/cows/game/roundSimulation/RoundSimulator.kt:25]: Here you can use map instead

`val units = attackInstruction.mapIndexed { index, unit -> UnitSimulationModel(unit.id, 1, 0, 2, 0, index * 2) }`

File: client/core/src/com/cows/game/roundSimulation/RoundSimulator.kt [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        return null
+    }
+
+    private fun calculateTowerAction(tower : TowerSimulationModel, units : MutableList<UnitSimulationModel>): SimulationAction? {
+        // if no target, either set new target if a unit is in range, or do nothing if no unit in range
+        if (tower.target == null) {
+            val newTarget = tower.findNewTarget(units) ?: return null
+            return TargetSimulationAction(tower, newTarget)
+        }
+        println(!tower.targetInRange())
```
Comment (AndreasWintherMoen) [inline client/core/src/com/cows/game/roundSimulation/RoundSimulator.kt:101]: Unnecessary println

File: client/core/src/com/cows/game/roundSimulation/simulationModels/SimulationAction.kt [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    }
+    override val type = ActionType.TARGET
+    override fun toJsonAction(): JsonAction {
+        return JsonAction(tower.id, type, unit?.id)
+    }
+
+}
+
+data class AttackSimulationAction(val tower: TowerSimulationModel): SimulationAction() {
+    override fun processAction() {
```
Comment (AndreasWintherMoen) [inline client/core/src/com/cows/game/roundSimulation/simulationModels/SimulationAction.kt:36]: Should be oneliner for consistency.

File: client/core/src/com/cows/game/roundSimulation/simulationModels/TowerSimulationModel.kt [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+            cooldown--
+        }
+    }
+
+    private fun setCooldown() {
+        cooldown = timeBetweenAttacks
+    }
+
+    fun attack() {
+        target!!.damage(damage)
```
Comment (AndreasWintherMoen) [inline client/core/src/com/cows/game/roundSimulation/simulationModels/TowerSimulationModel.kt:23]: Is target guaranteed to be not null?

Comment (AndreasWintherMoen) [reply client/core/src/com/cows/game/roundSimulation/simulationModels/TowerSimulationModel.kt:23]: You can use the _let_ syntax like in the _targetInRange_ method.

File: client/core/src/com/cows/game/roundSimulation/simulationModels/TowerSimulationModel.kt [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    fun targetInRange(): Boolean {
+        target?.let {
+            if (it.isDead) return false
+            return pathIndicesInRange.contains(it.pathIndex)
+        }
+        return false
+    }
+
+
+    fun findPathIndicesInRange(range: Int, path: List<IntArray>): List<Int> = path
```
Comment (AndreasWintherMoen) [inline client/core/src/com/cows/game/roundSimulation/simulationModels/TowerSimulationModel.kt:36]: sexy

File: client/core/src/com/cows/game/roundSimulation/simulationModels/UnitSimulationModel.kt [hunk 6]
```diff
@@ -0,0 +1,41 @@
+package com.cows.game.roundSimulation.simulationModels
+
+
+class UnitSimulationModel(
```
Comment (AndreasWintherMoen) [inline client/core/src/com/cows/game/roundSimulation/simulationModels/UnitSimulationModel.kt:4]: Some of these values, namely movementProgress, isSpawned and isDead don't need to be set in the constructor, as they should always have 0, false, false respectively as initial values

**Review by AndreasWintherMoen — CHANGES_REQUESTED** [review]
I've made some quick remarks. It's obvious WIP still, so I've ignored some excessive comments and print statements and just focused on the most egregious stuff.

File: client/core/src/com/cows/game/roundSimulation/RoundSimulator.kt (hunk 1 above)
Comment (Nicolaad) [reply client/core/src/com/cows/game/roundSimulation/RoundSimulator.kt:25]: Good point! I will also do that for the towers:)

File: client/core/src/com/cows/game/roundSimulation/simulationModels/TowerSimulationModel.kt (hunk 4 above)
Comment (Nicolaad) [reply client/core/src/com/cows/game/roundSimulation/simulationModels/TowerSimulationModel.kt:23]: It is in theory guaranteed at an earlier stage, but I agree that it should be a let statement

**Comment by Nicolaad** [comment]
The feedback should be resolved in 049030c , with some additional minor improvements:)

**Comment by Nicolaad** [comment]
I think this should now be out of WIP :)

**Review by AndreasWintherMoen — APPROVED** [review]
I think it looks good now. Let's get it merged in and we'll make the SS later.

**Pull request merged**